src/lib/polygon.js

```diff
@@ -0,0 +1,240 @@
+/**
+* Copyright 2012-2015, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+var dot = require('./matrix').dot;
+
+/**
+ * Turn an array of [x, y] pairs into a polygon object
+ * that can test if points are inside it
+ *
+ * @param ptsIn Array of [x, y] pairs
+ *
+ * @returns polygon Object {xmin, xmax, ymin, ymax, pts, contains}
+ *      (x|y)(min|max) are the bounding rect of the polygon
+ *      pts is the original array, with the first pair repeated at the end
+ *      contains is a function: (pt, omitFirstEdge)
+ *          pt is the [x, y] pair to test
+ *          omitFirstEdge truthy means points exactly on the first edge don't
+ *              count. This is for use adding one polygon to another so we
+ *              don't double-count the edge where they meet.
+ *          returns boolean: is pt inside the polygon (including on its edges)
+ */
+var polygon = module.exports = {};
+
+polygon.tester = function tester(ptsIn) {
+    var pts = ptsIn.slice(),
+        xmin = pts[0][0],
+        xmax = xmin,
+        ymin = pts[0][1],
+        ymax = ymin;
+
+    pts.push(pts[0]);
+    for(var i = 1; i < pts.length; i++) {
+        xmin = Math.min(xmin, pts[i][0]);
+        xmax = Math.max(xmax, pts[i][0]);
+        ymin = Math.min(ymin, pts[i][1]);
+        ymax = Math.max(ymax, pts[i][1]);
+    }
+
+    // do we have a rectangle? Handle this here, so we can use the same
+    // tester for the rectangular case without sacrificing speed
```

Comment:
Initially I thought I'd make separate testers for lasso and rectangle, but it seemed easier to just branch in here and attach a simplified `contains` function. There still may be cases in other trace types where you could improve rectangle performance by handling it explicitly - like for heatmap/contour - in that case the `selectPoints` routine can look at `poly.isRect` and `poly.xmin` etc.